### PR TITLE
Add LambdaCallback

### LINKED ISSUE #5343: Add LambdaCallback
## 🚀 Feature

Add `LambdaCallback` for quick and easy callback creation. 

### Motivation

Previously, we had to make a callback class in order to make a callback. However, Lambda Callback makes it easy to create a simple Callback without declaring a Class. TensorFlow officially provides this callback, which can be used very efficiently.

https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/LambdaCallback

### Pitch

Add a new callback, `LambdaCallback`

### PR DESCRIPTION
**Pull request by marload**
## What does this PR do?

Fixes #5343 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes [if needed]?
- [x] Did you write any new necessary tests [no need for typos, docs]? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by marload** [comment]
@tchaton

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        listeners = [m for m in dir(Callback) if not m.startswith("_")]
+        for k, v in kwargs.items():
+            if k not in listeners:
+                raise ValueError(f"Invalid argument `{k}`")
```
Comment (tchaton) [inline pytorch_lightning/callbacks/lambda_cb.py:44]:
```suggestion
            if k not in listeners:
                raise MisConfigurationException(f"The event function: `{k}` doesn't exist in supported callbacks function. Currently, Callback implements the following functions {dir(Callback)}")
```

Comment (tchaton) [reply pytorch_lightning/callbacks/lambda_cb.py:44]: Please, format this :)

**Review by tchaton — COMMENTED** [review]
Great work ! Could you update the raise error by providing a better message.
Can you make sure it works with all callback functions (auto-generate them and assert they are being called)

**Comment by SkafteNicki** [comment]
@marload could you please rebase on the `release/1.2-dev` branch as this PR right now contains a lot of changes from `master`

**Comment by tchaton** [comment]
Hey @marload,

Would you mind creating this PR from `release/1.2-dev` branch which is the feature branch ?

Best regards,
T.C

**Comment by Borda** [comment]
> Would you mind creating this PR from `release/1.2-dev` branch which is the feature branch ?

I have fixed it so no action needed

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+from pytorch_lightning.callbacks.base import Callback
+
+
+class LambdaCallback(Callback):
+    r"""
+    Create a simple callback on the fly.
+
+    Args:
+        **kwargs: event listener supported by ``Callback``
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/lambda_cb.py:31]: In lightning we usually refer to them as hooks
```suggestion
        **kwargs: hooks supported by ``Callback``
```

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    Example::
+
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        listeners = [m for m in dir(Callback) if not m.startswith("_")]
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/lambda_cb.py:41]:
```suggestion
        hooks = [m for m in dir(Callback) if not m.startswith("_")]
```

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        for k, v in kwargs.items():
+            if k not in listeners:
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/lambda_cb.py:43]:
```suggestion
            if k not in hooks:
```

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        **kwargs: event listener supported by ``Callback``
+
+    Example::
+
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/lambda_cb.py:40]: Can we make this more explicit? I.e. name every hook we support here? Otherwise people may not know what is supported and autocomplete won't be of any help here.

I know, that this increases the maintenance effort, but I think that this is essential for intuitive usage.

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# Copyright The PyTorch Lightning team.
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/lambda_cb.py:1]: should the file be renamed to `lambda.py` to be consistent with the other callbacks?

File: pytorch_lightning/callbacks/lambda_cb.py (hunk 5 above)
Comment (marload) [reply pytorch_lightning/callbacks/lambda_cb.py:40]: You are right
If this callback supports auto-completion, it will cost more to maintain.  tensorflow supports autocomplete, but they have far fewer hooks than we do.  I'm curious about the opinions of other maintainers.

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Lambda Callback
+^^^^^^^^^^^^^^^
+
+Create a simple callback on the fly.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:19]:
```suggestion
Create a simple callback on the fly using lambda functions.
```

File: pytorch_lightning/callbacks/__init__.py [hunk 8]
```diff
@@ -29,4 +29,5 @@
     'ModelCheckpoint',
     'ProgressBar',
     'ProgressBarBase',
+    'LambdaCallback',
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/__init__.py:32]: in sorted order pls?

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        hooks = [m for m in dir(Callback) if not m.startswith("_")]
+        for k, v in kwargs.items():
+            if k not in hooks:
+                raise ValueError(f"Invalid argument `{k}`")
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:44]: raise MisconfigurationException?

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    Example::
+
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        hooks = [m for m in dir(Callback) if not m.startswith("_")]
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:41]: maybe use `inspect`?
```python
import inspect
hooks = [m for m, _ in inspect.getmembers(Callback, predicate=inspect.isfunction)]
```

File: tests/callbacks/test_lambda_cb.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    test_callback = LambdaCallback(**callback_dicts)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_val_batches=1,
+        limit_train_batches=3,
+        limit_test_batches=2,
+        progress_bar_refresh_rate=0,
+        enable_pl_optimizer=True,
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:35]:
```suggestion
        fast_dev_run=True
```
`fast_dev_run` is all you need :)

File: tests/callbacks/test_lambda_cb.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+from tests.base.boring_model import BoringModel
+
+
+def test_lambda_call(tmpdir):
+    seed_everything(42)
+
+    model = BoringModel()
+    checker = set()
+
+    callback_dicts = {"setup": lambda *args: checker.add("setup")}
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:25]: maybe add at least one hook each from train/val/test??

**Comment by marload** [comment]
Could you please check the failed test? It doesn't seem to be the error I'm having in my PR.

**Comment by rohitgr7** [comment]
failing tests are being fixed here I believe https://github.com/PyTorchLightning/pytorch-lightning/pull/5352

File: docs/source/callbacks.rst [hunk 13]
```diff
@@ -102,6 +102,7 @@ Lightning has a few built-in callbacks.
     ModelCheckpoint
     ProgressBar
     ProgressBarBase
+    LambdaCallback
```
Comment (rohitgr7) [inline docs/source/callbacks.rst:105]: sorted order pls

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        hooks = [m for m, _ in inspect.getmembers(Callback, predicate=inspect.isfunction)]
+        for k, v in kwargs.items():
+            if k not in hooks:
+                raise MisconfigurationException(
+                    f"The event function: `{k}` doesn't exist in supported callbacks function."
+                    f"Currently, Callback implements the following functions {hooks}"
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:49]:
```suggestion
                    f" Currently, `Callback` implements the following functions {hooks}"
```

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LambdaCallback
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        hooks = [m for m, _ in inspect.getmembers(Callback, predicate=inspect.isfunction)]
+        for k, v in kwargs.items():
+            if k not in hooks:
+                raise MisconfigurationException(
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:47]: mind add a test raising this exception?

File: tests/callbacks/test_lambda_cb.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    hooks_args = {h: (lambda x: lambda *args: checker.add(x))(h) for h in hooks}
+    test_callback = LambdaCallback(**hooks_args)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, callbacks=[test_callback])
+
+    trainer.fit(model)
+    trainer.test(model)
+
+    print(len(hooks))
+    print(len(checker))
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:63]:
```suggestion
```

File: tests/callbacks/test_lambda_cb.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+    for h in hooks:
+        assert h in checker
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:66]:
```suggestion
    assert checker == set(hooks)
```

File: tests/callbacks/test_lambda_cb.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        "on_test_end",
+    ]
+    model = BoringModel()
+
+    hooks_args = {h: (lambda x: lambda *args: checker.add(x))(h) for h in hooks}
+    test_callback = LambdaCallback(**hooks_args)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, callbacks=[test_callback])
+
+    trainer.fit(model)
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:59]:
```suggestion
    results = trainer.fit(model)
    assert results
```

File: tests/callbacks/test_lambda_cb.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.base.boring_model import BoringModel
+
+
+def test_lambda_raise_misconfiguration():
+    try:
+        LambdaCallback(invalid=lambda *args: args)
+        assert False
+    except MisconfigurationException:
+        assert True
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:25]: use `pytest.raises`, you can check other tests for reference where it's used.

**Comment by Borda** [comment]
@marload feat-1.2 shall be fixed now, pls update your PR ;]

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class LambdaCallback(Callback):
+    r"""
+    Create a simple callback on the fly using lambda functions.
+
+    Args:
+        **kwargs: hooks supported by ``Callback``
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:34]:
```suggestion
        **kwargs: hooks supported by :class:`~pytorch_lightning.callbacks.base.Callback`
```

File: pytorch_lightning/callbacks/lambda_cb.py [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+        >>> trainer = Trainer(callbacks=[LambdaCallback(setup=lambda *args: print('setup'))])
+    """
+
+    def __init__(self, **kwargs):
+        hooks = [m for m, _ in inspect.getmembers(Callback, predicate=inspect.isfunction)]
+        for k, v in kwargs.items():
+            if k not in hooks:
+                raise MisconfigurationException(
+                    f"The event function: `{k}` doesn't exist in supported callbacks function."
+                    f"Currently, `Callback` implements the following functions {hooks}"
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lambda_cb.py:49]:
```suggestion
                    f"The event function: `{k}` does not exist in supported callbacks function."
                    f" Currently, `Callback` implements the following functions {hooks}"
```

File: tests/callbacks/test_lambda_cb.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+import pytest
+
+from pytorch_lightning import Trainer, seed_everything
+from pytorch_lightning.callbacks import LambdaCallback
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.base.boring_model import BoringModel
+
+
+def test_lambda_raise_misconfiguration():
+    with pytest.raises(MisconfigurationException):
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py:23]:
```suggestion
    with pytest.raises(MisconfigurationException, match='does not exist in supported callbacks function'):
```

**Review by rohitgr7 — APPROVED** [review]
minor updates. Also, rename the files pls `lambda.py` and `test_lambda.py`

File: pytorch_lightning/callbacks/lambda_cb.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/lambda_cb.py:43]: I agree with Justus here. We can write a test that will fail if someone adds a new callback method and does not update the LambdaCallback init (or renames, or deleteds callback methods).

File: tests/callbacks/test_lambda_cb.py [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+        "on_test_batch_end",
+        "on_train_start",
+        "on_train_end",
+        "on_test_start",
+        "on_test_end",
+    ]
+    model = BoringModel()
+
+    hooks_args = {h: (lambda x: lambda *args: checker.add(x))(h) for h in hooks}
+    test_callback = LambdaCallback(**hooks_args)
```
Comment (tchaton) [inline tests/callbacks/test_lambda_cb.py:55]: Would you mind adding this check ?

Check LambdaCallback.__init__ signature functions == Base class Callback hooks.  

So, if a new person add a callback hook to Callback, this will break the test and he will have to update  LambdaCallback.

Best,
T.C

Comment (marload) [reply tests/callbacks/test_lambda_cb.py:30]: Complete 😄

**Comment by marload** [comment]
I should test the `on_keyboard_interrrup` and `on_load_checkpoint` hooks, but i dont know what to do. Who can help me?

**Review by tchaton — APPROVED** [review]

**Comment by tchaton** [comment]
Hey @marload ,

You have 1 est failing. Can you resolve it ?

File: tests/callbacks/test_lambda_cb.py [hunk 24]
```diff
@@ -15,15 +15,9 @@
...
 from pytorch_lightning.callbacks import LambdaCallback
-from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from tests.base.boring_model import BoringModel
 
 
-def test_lambda_raise_misconfiguration():
-    with pytest.raises(MisconfigurationException, match='does not exist in supported callbacks function'):
-        LambdaCallback(invalid=lambda *args: args)
-
-
```
Comment (rohitgr7) [inline tests/callbacks/test_lambda_cb.py]: why did you remove this check?

**Review by rohitgr7 — COMMENTED** [review]
@marload `on_load_checkpoint` was a bit tricky since to trigger that it requires `on_save_checkpoint` to return something. So tried something. Mind check?

File: tests/callbacks/test_lambda_cb.py (hunk 24 above)
Comment (marload) [reply tests/callbacks/test_lambda_cb.py]: I added it again 👍

File: pytorch_lightning/callbacks/lambda_cb.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/lambda_cb.py:1]: +1 for renaming to lambda.py and test_lambda.py

File: tests/callbacks/test_lambda_cb.py (hunk 24 above)
Comment (marload) [reply tests/callbacks/test_lambda_cb.py]: Because `MisconfigurationException` no longer raise because `__init__` receives `Callable` objects as a real argument rather that `*kwargs`

File: docs/source/callbacks.rst [hunk 25]
```diff
@@ -98,6 +98,7 @@ Lightning has a few built-in callbacks.
     EarlyStopping
```
Comment (rohitgr7) [inline docs/source/callbacks.rst:98]: @tchaton did you forget to include `BackboneLambdaFinetuningCallback` here, or is it on purpose?

**Review by rohitgr7 — APPROVED** [review]
again, can you please rename the files? else LGTM.

File: docs/source/callbacks.rst (hunk 25 above)
Comment (marload) [reply docs/source/callbacks.rst:98]: My IDE warn to me because `lambda` is python built-in identifier.
So, below code raise error.
```
from pytorch_lightning.callbacks.lambda import LambdaCallback
```
Therefore I cannot change file name from `lambda_cb.py` to `lambda.py` 😢

Comment (rohitgr7) [reply docs/source/callbacks.rst:98]: hmm... I use vim, so can't say what to do here. cc @awaelchli @SkafteNicki

Comment (awaelchli) [reply docs/source/callbacks.rst:98]: yes, it seems this built in name cannot be used in an import statement. It will be a syntax error.

```bash
  File "C:\Users\aeduw\Documents\Repositories\pytorch-lightning\pytorch_lightning\callbacks\__init__.py", line 21
    from pytorch_lightning.callbacks.lambda import LambdaCallback
                                          ^
SyntaxError: invalid syntax

```
No way around it, we cannot name it lambda.py :( 
Could consider name it `lambda_function.py`
or just misspelling it `lamda.py` 🤣

Comment (awaelchli) [reply docs/source/callbacks.rst:98]: Interestingly, we can have a file print.py and import it without any problem :)

Comment (marload) [reply docs/source/callbacks.rst:98]: I think 'lambda_function.py' would be betther than 'lamda.my' 😀 
I will update it

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

File: docs/source/callbacks.rst (hunk 25 above)
Comment (marload) [reply docs/source/callbacks.rst:98]: I updated it 👍

**Review by awaelchli — APPROVED** [review]
